[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73991
* #73396

Automatically generate `datapipe.pyi` via CMake and removing the generated .pyi file from Git. Users should have the .pyi file locally after building for the first time.

I will also be adding an internal equivalent diff for buck.

Differential Revision: [D34868001](https://our.internmc.facebook.com/intern/diff/D34868001)

NivekT committed:
[DataPipe] Automatically generate datapipe.pyi via CMake

[ghstack-poisoned]

albanD reviewed (approved):
The part about the build system update LGTM

You will most likely have to update your script to run fine from where it runs now (relative imports do help for these).

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake

Since it will be automatically generated, I believe we should remove the output file from git VCS (like other generated .pyi file).

[ghstack-poisoned]

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake

Since it will be automatically generated, I believe we should remove the output file from git VCS (like other generated .pyi file).

[ghstack-poisoned]

NivekT commented:
> The part about the build system update LGTM
> 
> You will most likely have to update your script to run fine from where it runs now (relative imports do help for these).

I removed a dependency and I think it should work fine now

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake

Since it will be automatically generated, I believe we should remove the output file from git VCS (like other generated .pyi file).

[ghstack-poisoned]

NivekT commented:
@ejguan @albanD Should I be removing the generated .pyi file from git VCS? 

Here is the trade off (should only impact PyTorch developers):
1. Remove the file from Git - the .pyi will need be generated by running `python setup.py develop` or similar **every time** you checkout a new branch
2. Keep the file in Git - when you go to a new branch, the Git version of the .pyi file will exist. When you run `python setup.py develop` and that will overwrite the Git version with the generated version. The Git version might be out of sync (generated version will overwrite the Git version). 
    - We keep the file, should we add it to `.gitignore` or not? 

I am currently keeping the .pyi file in Git but adding it to `.gitignore`.

ejguan commented:
> Remove the file from Git - the .pyi will need be generated by running python setup.py develop or similar every time you checkout a new branch

It's annoying because functional API is pure python rather than the binding from C++ to Python. That's why I prefer generating pyi file whenever `functional_datapipe` is invoked. Even new DataPipe is implemented, the pyi file should be updated at that time. Then, even though users are using `python setup.py develop`, they should be able to use these APIs.

> I am currently keeping the .pyi file in Git but adding it to `.gitignore`

If you want to generate pyi via CMake, this is better.

albanD commented:
> Remove the file from Git - the .pyi will need be generated by running python setup.py develop or similar every time you checkout a new branch

Not sure to follow. My understanding if the file is not in git and is in gitignore: whatever was there before you change branch will remain there.
So yes, after the first clone you will have to do a compilation to get the file.
And when you change branch (or modify python code locally) that pyi file might get stale until you recompile.

If my understanding is correct then removing it and ignoring it sounds like the right move. Having a file here will potentially lead to conflicts. And a stale pyi is not a big issue during development.

albanD reviewed (changes requested):
Sorry for coming back on the review but following the discussions here I think we want to actually remove `torch/utils/data/datapipe.pyi` from the repo. The same way it is done for `torch/_C/_VariableFunctions.pyi`.

ejguan commented on torch/CMakeLists.txt:
```diff
@@ -196,14 +196,18 @@ add_custom_command(
+    "${TORCH_SRC_DIR}/utils/data/datapipe.pyi"
     COMMAND
-    "${PYTHON_EXECUTABLE}" -mtools.pyi.gen_pyi
-      --native-functions-path "aten/src/ATen/native/native_functions.yaml"
-      --deprecated-functions-path "tools/autograd/deprecated.yaml"
+      "${PYTHON_EXECUTABLE}" -mtools.pyi.gen_pyi
+        --native-functions-path "aten/src/ATen/native/native_functions.yaml"
+        --deprecated-functions-path "tools/autograd/deprecated.yaml"
+    COMMAND
+      "${PYTHON_EXECUTABLE}" ${TORCH_SRC_DIR}/utils/data/gen_pyi.py
```
Not an expert on CMake. Should we separate it into a new `add_custom_command`?

NivekT replied:
Agree

NivekT commented:
> Sorry for coming back on the review but following the discussions here I think we want to actually remove `torch/utils/data/datapipe.pyi` from the repo. The same way it is done for `torch/_C/_VariableFunctions.pyi`.

I will be removing that and also add an internal diff for buck

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake and removing the generated .pyi file from Git. Users should have the .pyi file locally after building for the first time.

I will also be adding an internal equivalent diff for buck.


[ghstack-poisoned]

NivekT commented on torch/CMakeLists.txt:
```diff
@@ -210,6 +210,18 @@ add_custom_command(
     "${TORCH_ROOT}"
 )
+file(GLOB_RECURSE datapipe_files "${TORCH_SRC_DIR}/utils/data/datapipes/*.py")
+add_custom_command(
```
Hi @malfet, I added this new custom command here but it doesn't seem to be running when I run `python setup.py develop`. Any idea why that might be?

I tried a few things: should I wrap double quote around `${datapipe_files}`? Or maybe I'm using `GLOB_RECURSE` or its output improperly?

cc: @albanD

NivekT replied:
I went through this again. Adding the output file as a dependency on L193 will make it run as needed.

This is now working as intended.

NivekT commented on torch/CMakeLists.txt:
```diff
@@ -210,6 +210,18 @@ add_custom_command(
+    "${PYTHON_EXECUTABLE}" ${TORCH_SRC_DIR}/utils/data/gen_pyi.py
+    DEPENDS
+    "${TORCH_SRC_DIR}/utils/data/datapipe.pyi.in"
+    ${datapipe_files}
```
Yea, the output .pyi file depends on the content of the DataPipe files.

ejguan replied:
Yeah, understood. Just curious if this command works without adding DataPipe files as dependency.

NivekT replied:
Turns out this wasn't the issue. The change on L193 is what's needed.

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake and removing the generated .pyi file from Git. Users should have the .pyi file locally after building for the first time.

I will also be adding an internal equivalent diff for buck.


[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34868001).

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake and removing the generated .pyi file from Git. Users should have the .pyi file locally after building for the first time.

I will also be adding an internal equivalent diff for buck.

Differential Revision: [D34868001](https://our.internmc.facebook.com/intern/diff/D34868001)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34868001).

ejguan reviewed (approved):
LGTM

NivekT committed:
Update on "[BE Hackathon][DataPipe] Automatically generate datapipe.pyi via CMake"


Automatically generate `datapipe.pyi` via CMake and removing the generated .pyi file from Git. Users should have the .pyi file locally after building for the first time.

I will also be adding an internal equivalent diff for buck.

Differential Revision: [D34868001](https://our.internmc.facebook.com/intern/diff/D34868001)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D34868001).
